### DISCUSSION TITLE
Can't get geolocation to work with events API

### DISCUSSION THREAD

**Discussion by jeff-h** [plausible/analytics#2767] [Self-Hosted Support] [answered]
I have Plausible up and running perfectly for my sites, with the Regions/geolocation feature working perfectly. I'm now looking at the [events API](https://plausible.io/docs/events-api) for in-app tracking (for my desktop Mac app). Using CURL to play with this is all working just fine also, with the exception of the geo / regions feature. Everything is recorded as originating from my nearest cloudflare cache server (I think).

My setup is like this:
VPS: Docker running Plausible
Cloudflare: using their DNS to point my plausible domain name to the VPS

When I make a CURL request to my Plausible instance, it all works except the location is ignoring the `X-Forwarded-For`:

```
curl -X "POST" "https://my-plausible-server.com/api/event" \
     -H 'User-Agent: MyAppName/1.0.0' \
     -H 'X-Forwarded-For: 151.101.0.81' \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
                  "name": "pageview",
                  "url": "https://my-app/test",
                  "domain": "my-app"
        }'
```

You'll notice I have set X-Forwarded-For to `151.101.0.81`, which is a randomly-chosen IP in the US, just to see if my Plausible geolocates the request there. It doesn't, and instead says a region near me in Au.

Unfortunately nothing I do seems to be able to get Plausible to honour the `X-Forwarded-For` header that I'm sending it. Am I overlooking something obvious? Can anyone think of something I could try in order to debug this?

**Comment by ruslandoga** [comment, marked as answer]
👋 @jeff-h 

Note that [`cf-connecting-ip`](https://developers.cloudflare.com/fundamentals/get-started/reference/http-request-headers/#cf-connecting-ip) header [takes precedence](https://github.com/plausible/analytics/blob/d0895efbddb0e4e3e6e0d36bf97c5b43302e2fca/lib/plausible_web/remote_ip.ex#L9-L10) over `x-forwarded-for` header in Plausible. So I think when proxying through CloudFlare, `-H 'X-Forwarded-For: 151.101.0.81'` is practically ignored.

**Comment by jeff-h** [comment]
Wow, thanks @ruslandoga — really appreciate your input. It totally explains why my `X-Forwarded-For` has failed thus far no matter what I did!

Based on your input I went into Cloudflare and attempted to add a "transform rule" to remove the `CF-Connecting-IP` but got the following:

`'remove' is not a valid value for operation because it cannot be used on header beginning with 'cf-' (Code: 20087)`

So it seems Cloudflare don't allow you to remove it :(

Do you think Plausible would be open to the idea of a custom header (say, `plausible-client-ip`) which if present would take precedence over any others? This would allow an app to definitively provide Plausible with the client ip regardless of what else may be involved (Cloudflare in this case).

**Reply by ruslandoga** [reply]
Why do you want to remove `CF-Connecting-IP`? According to CloudFlare docs, it should contain the correct originating IP address of the client.

**Reply by jeff-h** [reply]
My original issue description was probably rather too distilled, in an effort to isolate the actual issue.

My setup involves a Mac app which calls a Laravel app (but not for analytics purposes). Laravel thus knows the client IP (ie from the Mac running the app). Using this, when certain things happen Laravel generates specific Plausible event calls, and needs Plausible to see the originating Mac's IP, not the VPS which Laravel is running on, despite the latter being where the event request originates.

So Laravel sets the `X-Forwarded-For` header to the Mac's IP, but Cloudflare's `CF-Connecting-IP` points stubbornly at the VPS's IP. When I remove the `CF-Connecting-IP` header, it seems CF is happy to use the `X-Forwarded-For` instead.

**Reply by visini** [reply]
~Awesome, removing the `CF-Connecting-IP` request header via the transform rule actually worked. Thank you!~

Edit: This stopped working a few days ago.

**Comment by jeff-h** [comment]
Hmm, I tried a "Modify Request Header" rule in Cloudlfare (Rules>Transform Rules) to remove the `CF-Connecting-IP` and it appears that may actually have worked. I suspect if you remove it on the incoming request, Cloudflare may still add it in the proxied request to the originating server (Plausible in this case), but perhaps basing it on the incoming `X-Forwarded-For`. Hard to say without seeing what Plausible is receiving.

Two things:
* I still greatly prefer the custom `plausible-client-ip` idea over configuring Cloudflare's dashboard 🙄
* it has been really hard to debug this issue; it would be awesome if Plausible had some kind of debug mode which would log incoming requests. I realise few would probably use this though. My only option was to try to log the requests in my Nginx reverse proxy but that was not easy and didn't feel quite definitive enough for me to trust it.

**Reply by ruslandoga** [reply]
You can connect to your running instance and print or trace the incoming requests

```console
$ cd hosting
$ docker compose exec plausible bin/plausible remote
```
and there you can add a tracer for the `RemoteIP` module to print the incoming headers
```elixir
iex> headers = fn {:trace, _pid, :call, {_mod, _fun, [%Plug.Conn{req_headers: headers}]}} -> inspect(headers) end
iex> mod_fun_args = {PlausibleWeb.RemoteIP, :get, 1}
iex> how_many_times = 5
iex> :recon_trace.calls(mod_fun_args, how_many_times, formatter: headers)
```
Example output for:
```sh
curl http://localhost:8000/api/event -d '{"name": "pageview","url":"https://dummy.site/baldur","domain":"dummy.site","referrer":null,"screen_width":520}' -H 'content-type: application/json'
```
```
[{"accept", "*/*"}, {"content-length", "111"}, {"content-type", "application/json"}, {"host", "localhost:8000"}, {"user-agent", "curl/7.86.0"}]
```

**Reply by jeff-h** [reply]
Wow, that is awesome! Thanks for letting me know.

Re: the original issue, I pushed my stuff to production and can confirm that the Cloudflare config definitely has fixed the issue of geolocation (ie getting the correct client IP to Plausible).
